plague (Spyder Bite)

JaskRendix opened the pull request:
PR introduce the mechanics for plague (Spyder Bite).
- creates a strenum inside db.py (PlagueType: sickless, inoculated and infected);
- creates a field plague in npc.py (PlagueType.sickless) -> as default;
- creates a field plague in monster.py (PlagueType.sickless) -> as default -> the field will be saved;
- creates new condition "party_infected" to check how many infected monsters inside the party;
- creates new action "npc_plague" to set the NPC as infected;
- creates new action "quarantine" to move all the infected monsters inside the box (in) or outside the box (out) - (event Spyder campaign - Candy Town);
- creates new action "set_monster_plague" to heal all the infected monster (or infect all, or inoculate all);

tested, isort, black, no new typehints

@Sanglorian  

- Do not display the icon indicating that a monster has Spyder Bite -> done
- It is not healed by a “full heal”. -> done
- A monster can have other conditions as well as Spyder Bite. -> done
- Even being eliminated does not cause a monster to lose Spyder Bite. -> done
- Some enemy trainers have the property “infected”, which means that their monsters have Spyder Bite. -> done
- Some enemy trainers have the property “inoculated”, which means that their monsters cannot be afflicted by Spyder Bite. -> done
- There will be a special location in the game that cures monsters that have Spyder Bite -> where?
- combat mechanism -> done

Questions:
- if a monster evolves? If rockitten is infected, rockat is infected?
- all the trainers in the wayfarer inn are infected? other places?

JaskRendix committed:
plague

JaskRendix committed:
Merge branch 'development' into plague

Sanglorian commented:
So good, thanks! 

* There will be a special location in the game that cures monsters that have Spyder Bite -> where?

This is the Hospital. In the plan, Just interacting with "The cure!" event should cure all tuxemon and stop the player's tuxemon from becoming infected in future. (In the plan, the Hospital is a whole dungeon with different challenges and trainers to battle, but it'll still be an interaction in the hospital that serves as the cure). 

* if a monster evolves? If rockitten is infected, rockat is infected?

Yes, they should stay infected in their evolved form. 

* all the trainers in the wayfarer inn are infected? other places?

Good question! At the moment, let's keep it limited to the wayfarer inn as that's where the sick have been sent. Maybe there should be random trainers beyond the wayfarer inn who are also infected who haven't been detected yet.

JaskRendix commented:
> Good question! At the moment, let's keep it limited to the wayfarer inn as that's where the sick have been sent. Maybe there should be random trainers beyond the wayfarer inn who are also infected who haven't been detected yet.

Wayfarer_inn1 and 2 are done.
One question: this [guy](https://wiki.tuxemon.org/Wayfarer_Inn#Spare) , the spare, the maniac who wants to give the bot bot. Is he supposed to give the monster without choice (Y/N)? I'm asking because I found it a bit weird, if you want I can add the Y/N event.

Think about it, if you decide for Y/N we could infect all the monsters in the party in that moment, but at least we give the choice to the player.

> Yes, they should stay infected in their evolved form.

Done.

> This is the Hospital. In the plan, Just interacting with "The cure!" event should cure all tuxemon and stop the player's tuxemon from becoming infected in future.

Do you mean the monster in the player's party?
What about a monster infected inside a box?
I guess since the player is inoculated, each monster infected withdrawn from the box is cured and inoculated, right?

Sanglorian commented:
> Wayfarer_inn1 and 2 are done. One question: this [guy](https://wiki.tuxemon.org/Wayfarer_Inn#Spare) , the spare, the maniac who wants to give the bot bot. Is he supposed to give the monster without choice (Y/N)? I'm asking because I found it a bit weird, if you want I can add the Y/N event.
> 
> Think about it, if you decide for Y/N we could infect all the monsters in the party in that moment, but at least we give the choice to the player.

That's a good point - yes, please make it a choice. And I like the idea that doing so immediately infects all monsters in the party at that moment. 

> Do you mean the monster in the player's party? What about a monster infected inside a box? I guess since the player is inoculated, each monster infected withdrawn from the box is cured and inoculated, right?

Yeah, at that point even infected monsters in a box should be cured. 

Although I'd actually like to make it so infected monsters can't be boxed, if possible. Instead an error message could play: "Virus detected. For the safety of all tuxemon in our system, you cannot store an infected tuxemon."

JaskRendix commented:
It sounds good.

Another question: the [confiscation](https://wiki.tuxemon.org/Candy_Town#1._Confiscation) must be triggered by the **seen_candy** event?
![Screenshot from 2023-04-14 17-21-58](https://user-images.githubusercontent.com/64643719/232086258-e0962d97-c5b5-41fe-b8f7-f46b42388539.png)
Or is it necessary to split it and make it one for the river and another for the path?

JaskRendix commented:
@Sanglorian It sounds good.

Another question: the [confiscation](https://wiki.tuxemon.org/Candy_Town#1._Confiscation) must be triggered by the **seen_candy** event?
`removed pictures for readability`
Or is it necessary to split it and make it one for the river and another for the path?

regarding this event:
> We are taking all infected tuxemon to the Hospital to be quarantined

I have already created the action to quarantine all the infected monster, these will be sent in a specific box called quarantine. The box label will be visible in the PC but not accessible (unable to drop or pick monsters). The action has an "in" and "out" function

So if you want to multiply the confiscation points with `quarantine in` (after wayfarer inn), feel free to tell me where do you want to put those. So you can create a specific setting (after the hospital), where we can `quarantine out` all of them (inoculated). Like another event (event to retake my monster).

Regarding this, just one point, since the max for box is 30 monsters, and forecasting 6 monsters infected each time, then you can set other 4 confiscation points. Total 5, 5*6 = 30, that's perfect. At the end the confiscation is an unique event not repeatable, right? The first in Timber or before, then following in successive maps.

---

> That's a good point - yes, please make it a choice. And I like the idea that doing so immediately infects all monsters in the party at that moment.

Done. This is now the botbot event in Wayfarer Inn.
```
msgid "spyder_wayfarer1_maniac"
msgstr "Hey, buddy, do you want my Botbot?"
msgid "spyder_wayfarer1_maniac1" ---> split from the one above
msgstr "I dunno how long I'm going to be stuck here, so it might as well get to explore with you."
msgid "spyder_wayfarer1_maniac2"
msgstr "A little privacy, please! " --> follow up for both accepting or refusing the botbot
```
eventually if you got inspired we can add a different output if the player refuses or accepts, but _A little privacy, please!_ can be a good way of ending both actions.

---

> Although I'd actually like to make it so infected monsters can't be boxed, if possible. Instead an error message could play: "Virus detected. For the safety of all tuxemon in our system, you cannot store an infected tuxemon."

Done.
`removed pictures for readability`

---

> This is the Hospital. In the plan, Just interacting with "The cure!" event should cure all tuxemon and stop the player's tuxemon from becoming infected in future.

Done.

JaskRendix committed:
plague

Sanglorian commented:
Thanks so much! 

> Another question: the [confiscation](https://wiki.tuxemon.org/Candy_Town#1._Confiscation) must be triggered by the **seen_candy** event? ![Screenshot from 2023-04-14 17-21-58](https://user-images.githubusercontent.com/64643719/232086258-e0962d97-c5b5-41fe-b8f7-f46b42388539.png) Or is it necessary to split it and make it one for the river and another for the path?

There should be a Gunner NPC blocking the land entrance from Route 6, so the player should enter by the river. So just the river event should be fine - maybe an Enforcer NPC could walk up to shore near the player? 

> regarding this event:
> 
> > We are taking all infected tuxemon to the Hospital to be quarantined
> 
> I have already created the action to quarantine all the infected monster, these will be sent in a specific box called quarantine. The box label will be visible in the PC but not accessible (unable to drop or pick monsters). The action has an "in" and "out" function
> 
> So if you want to multiply the confiscation points with `quarantine in` (after wayfarer inn), feel free to tell me where do you want to put those. So you can create a specific setting (after the hospital), where we can `quarantine out` all of them (inoculated). Like another event (event to retake my monster).
> 
> Regarding this, just one point, since the max for box is 30 monsters, and forecasting 6 monsters infected each time, then you can set other 4 confiscation points. Total 5, 5*6 = 30, that's perfect. At the end the confiscation is an unique event not repeatable, right? The first in Timber or before, then following in successive maps.

From memory, Candy Town's the only place where the confiscation happens, and it should be unique -- so I think we only need 6 slots maximum. But doesn't hurt to have the full 30. 

The player should get them all back after the Hospital event. I'd actually like to have the option to give them back one or more random monsters of theirs that were confiscated -- the story being that they find their quarantined monsters while they search through the hospital. Do you think that would be possible?

JaskRendix committed:
Merge branch 'Tuxemon:development' into plague

JaskRendix commented:
> Do you think that would be possible?

Yes, just updated the quarantine action, with quarantine out you can set the number of monsters:
`<property name="act10" value="quarantine out,5"/>`

let's say you have 30 monsters in the box, well, it'll choose the 5 by random, and it'll add those to the party, if the party is full, those monster will be deposited in the classic Kennel box. If you try to take out more than the elements inside, then it'll take out everything (like the example above, but there are only 4 monsters in the box).

That's why I asked above about creating other confiscation points (before the entrance). It depends if you want to have a minimum number of monster quarantined.
If you plan your hospital quest in 6 acts and you plan to give 1 monster at each act, then you need at least 6 monsters.

> There should be a Gunner NPC blocking the land entrance from Route 6, so the player should enter by the river. So just the river event should be fine - maybe an Enforcer NPC could walk up to shore near the player?

Event done. Used red knight sprite.
`removed pictures for readability`

Questions (some I forgot to ask in the Dojo PR):
1. is the Dojo a mandatory part of the story? (it means if I need to put a monk and block the access to the route5 or nimrod door);
2. is Billie infected? If not, do we need to set her monster as inoculated?
3. do you plan to insert a couple of fights with Billie? Like after the wayfarer inn and in route5/6?
4. is it wrong Flower Town right? I think you mean Timber City.
5. Nudiflot lv10: is inoculated or simply free from sickness?

`(If all tuxemon in the party are infected, then teleport Protagonist to Flower City Cafe) `
(https://wiki.tuxemon.org/Candy_Town#1._Confiscation)?

Sanglorian commented:
Thanks! 
> Questions (some I forgot to ask in the Dojo PR):
> 
>     1. is the Dojo a mandatory part of the story? (it means if I need to put a monk and block the access to the route5 or nimrod door);

No, all good - it's an optional side quest.
> 
>     2. is Billie infected? If not, do we need to set her monster as inoculated?

Good question. By Candy Town, Billie should be infected.
> 
>     3. do you plan to insert a couple of fights with Billie? Like after the wayfarer inn and in route5/6?

Yeah, good idea!
> 
>     4. is it wrong Flower Town right? I think you mean Timber City.
> 

I was wrong about Flower Town, but it should actually be Candy Town Cafe. I.e. the player does still get into town even though their tuxemon may be confiscated.
>     5. Nudiflot lv10: is inoculated or simply free from sickness?
> 

Good question - just free from sickness.

Hotel design looks good but I think it should be at least two floors, and more maze like. There could be quite a few trainer battles here - maybe 10? And I like the idea of the player getting a little lost, if possible.

I have some notes on potential NPCs and dialogues already if that would help.

JaskRendix commented:
> Yeah, good idea! 

if you have the dialogues (+ monsters and levels), we can put in these places

`removed pictures for readability`

then if you have other dialogues, we can put some NPCs here too, because someone will try to pass for here with the surfboard.
`removed pictures for readability`

> I was wrong about Flower Town, but it should actually be Candy Town Cafe. I.e. the player does still get into town even though their tuxemon may be confiscated.

Ok, just created the Café in Candy too. I had already created the event in Timber, but I moved it here.
`removed pictures for readability`

> Hotel design looks good but I think it should be at least two floors, and more maze like. There could be quite a few trainer battles here - maybe 10? And I like the idea of the player getting a little lost, if possible.

Sanglorian commented:
> > Yeah, good idea!
> 
> if you have the dialogues (+ monsters and levels), we can put in these places ![Screenshot from 2023-04-17 13-44-38](https://user-images.githubusercontent.com/64643719/232475413-378c4272-6196-46d8-810b-4fa6e2a1d2bb.png) ![Screenshot from 2023-04-17 13-43-38](https://user-images.githubusercontent.com/64643719/232475416-0b015299-4122-4b18-89fa-ee2b5d91a20c.png)

Route 4:
"I've got all the latest gear and the best tuxemon. What have you got?"
STARTER level 18, Cardiwing level 16, Eyesore level 16, Viviphyta level 16
"I guess money can't buy everything, after all."

Route 6: 

"Let's make it quick. I have a spa appointment in an hour."
STARTER level 34, Cardiwing level 30, Eyesore level 30, Viviphyta level 30
"Huh. You're a better trainer than I thought."

> 
> then if you have other dialogues, we can put some NPCs here too, because someone will try to pass for here with the surfboard. ![Screenshot from 2023-04-17 13-45-49](https://user-images.githubusercontent.com/64643719/232475500-de1a6f91-75f8-4b97-86b4-1904426b1674.png)

Actually would it be possible to block it with a rock for now? Just here --

![image](https://user-images.githubusercontent.com/1503938/232758650-eb1b3a55-f626-4212-a4cc-1d206f98bdb5.png)

> 
> > I was wrong about Flower Town, but it should actually be Candy Town Cafe. I.e. the player does still get into town even though their tuxemon may be confiscated.
> 
> Ok, just created the Café in Candy too. I had already created the event in Timber, but I moved it here. ![spyder_candy_cafe](https://user-images.githubusercontent.com/64643719/232479003-11595923-449a-4d20-b23a-449f037ed223.png)
> 

Thanks! 

> > Hotel design looks good but I think it should be at least two floors, and more maze like. There could be quite a few trainer battles here - maybe 10? And I like the idea of the player getting a little lost, if possible.
> 
> first and second floor, regarding the interior ![spyder_hospital](https://user-images.githubusercontent.com/64643719/232486678-b8429ba4-7616-413a-95cd-38c23e419505.png) - ![spyder_hospital2](https://user-images.githubusercontent.com/64643719/232486702-dbe85e6b-4edf-44e1-bf66-6648a60f7637.png) since there is the elevator we can do multiple similar floors

Looking good! 
 
Some ideas for NPCs in the hospitals: 

Spyder Rookie:
* You look like you could use a challenge.
* Anu level 30, Galnec level 30
* Well, I didn't say it would be me. 

Spyder Boss: 
* You will quickly learn that you have chosen for your sparring partner the most skilled trainer in the building.
* Exclawvate level 40
* Not too shabby. I don't suppose I could recruit you for Spyder?

Spyder Boss:
* Salutations. Test your puissance against my menagerie.
* Boltnu level 30, Taupypus level 30, Snaki level 30, Galnec level 30
* Bollocks.

Spyder Rookie
* Boss says we can't help ourselves to the stashed tuxemon ... yet.
* Anu level 30, Selket level 30
* This would have been a lot easier with a Moloch or two!

Spyder Boss
* You wouldn't believe the powerful mons these suckers have handed over. 
* Boltnu level 36, Snaki level 36
* You may have beaten me, but you'll never stop Spyder.

Spyder Rookie (no battle)
I don't recognize you from the induction.

Spyder Rookie:
* What happened to your uniform? Wait, imposter! Stop right there!
* Anu level 34, Taupypus level 27
* I'm going to have to report this to head office. Thanks for nothing. 

Spyder Boss (looking over a box full of tuxeballs or cage of tuxemon):
* Like taking candy from a baby.
* Galnec level 36, Exclawvate level 36
* Waa!

What's the point of having so many of us stationed here. No one will get through the DNA blocker.

Spyder Rookie:
* Oi! Stop right there!
* Selket level 32, Taupypus level 32
* You may have beaten me, but there's no way you'll beat what they have cooking at HQ.

Spyder Boss (no battle):
Confiscating all the tuxemon was a waste of time. No trainer can stop the super-monsters that Spyder is breeding. 

Scientist (no battle):
We have a cure, but we keep running into the Cathedral's DRM system.

Nurse (no battle): 
You have to stop them! These villains have taken over the hospital, and they're stopping us from releasing the cure.

JaskRendix commented:
Perfect, I'm going to split the hospital in another PR.

JaskRendix commented:
@Sanglorian perfect, I'm going to split the hospital in another PR. #1765

JaskRendix committed:
plague

JaskRendix committed:
Merge branch 'development' into plague

JaskRendix committed:
Merge branch 'development' into plague

JaskRendix commented:
@Sanglorian since we have split the following issues:
split #1765 hospital (work in progress);
split #1766 encounters with Billie (merged);
split #1732 block river with rock (merged);

is there something still missing regarding the plague? If not we are ready to merge. Let me know.

Sanglorian commented:
I think we're good to merge -- thank you so much!

kerizane commented on tuxemon/db.py:
```diff
@@ -44,6 +44,12 @@ class ItemSort(str, Enum):
 
 
+class PlagueType(str, Enum):
+    sickless = "sickless"
```
Can we call this "healthy"?

JaskRendix replied:
done

kerizane reviewed (changes requested):
Comment

JaskRendix committed:
plague
